docs/pages/release_notes.md

```diff
@@ -13,13 +13,71 @@ This is a minor release.
 ### Table Of Contents
 
 * [New and noteworthy](#new-and-noteworthy)
+    *   [Drawing a line between private and public API](#drawing-a-line-between-private-and-public-api)
     *   [New Rules](#new-rules)
 * [Fixed Issues](#fixed-issues)
 * [API Changes](#api-changes)
 * [External Contributions](#external-contributions)
 
 ### New and noteworthy
 
+#### Drawing a line between private and public API
+
+Until now, all released public members and types were implicitly considered part
+of PMD's public API, including inheritance-specific members (protected members, abstract methods).
+We have maintained those APIs with the goal to preserve full binary compatibility between minor releases,
+only breaking those APIs infrequently, for major releases.
+
+
+In order to allow PMD to move forward at a faster pace, this implicit contract will
+be invalidated with PMD 7.0.0. We now introduce more fine-grained distinctions between
+the type of compatibility support we guarantee for our libraries, and ways to make
+them explicit to clients of PMD.
+
+##### `.internal` packages and `@InternalApi` annotation
+
+*Internal API* is meant for use *only* by the main PMD codebase. Internal types and methods
+may be modified in any way, or even removed, at any time.
+
+Any API in a package that contains an `.internal` segment is considered internal.
+The `@InternalApi` annotation will be used for APIs that have to live outside of
+these packages, e.g. methods of a public type that shouldn't be used outside of PMD (again,
+these can be removed anytime).
+
+##### `@ReservedSubclassing`
```

Comment:
> One thing, we might change for 7.0.0, is: making the AST classes final and making any constructor of an abstract intermediate class package-private.

I think that would suffice to eliminate nearly all `@ReservedSubclassing` annotations at 7.0.0 time. The remaining sticky point is the AST interfaces, which we can't prevent from being extended, since they have no constructors. If we migrate our codebase to Java 8 we may rely even more on interfaces since they could be used as "mixins", which could be extremely useful to share code between nodes

I'd argue that we should keep it for these use cases, but this could be also done by some clear documentation, with a little more writing I guess.